source/extensions/filters/http/json_to_metadata/filter.cc

```diff
@@ -0,0 +1,391 @@
+
+namespace {
+
+struct JsonValueToStringConverter {
+  std::string operator()(bool&& val) { return std::to_string(val); }
+  std::string operator()(int64_t&& val) { return std::to_string(val); }
+  std::string operator()(double&& val) { return std::to_string(val); }
+  std::string operator()(std::string&& val) { return std::move(val); }
+};
+
+struct JsonValueToDoubleConverter {
+  absl::StatusOr<double> operator()(bool&& val) { return static_cast<double>(val); }
+  absl::StatusOr<double> operator()(int64_t&& val) { return static_cast<double>(val); }
+  absl::StatusOr<double> operator()(double&& val) { return val; }
+  absl::StatusOr<double> operator()(std::string&& val) {
+    double dval;
+    if (absl::SimpleAtod(StringUtil::trim(val), &dval)) {
+      return dval;
+    }
+    return absl::InternalError(fmt::format("value {} to number conversion failed", val));
+  }
+};
+
+struct JsonValueToProtobufValueConverter {
+  absl::StatusOr<ProtobufWkt::Value> operator()(bool&& val) {
+    ProtobufWkt::Value protobuf_value;
+    protobuf_value.set_bool_value(val);
+    return protobuf_value;
+  }
+  absl::StatusOr<ProtobufWkt::Value> operator()(int64_t&& val) {
+    ProtobufWkt::Value protobuf_value;
+    protobuf_value.set_number_value(val);
+    return protobuf_value;
+  }
+  absl::StatusOr<ProtobufWkt::Value> operator()(double&& val) {
+    ProtobufWkt::Value protobuf_value;
+    protobuf_value.set_number_value(val);
+    return protobuf_value;
+  }
+  absl::StatusOr<ProtobufWkt::Value> operator()(std::string&& val) {
+    if (val.size() > MAX_PAYLOAD_VALUE_LEN) {
+      return absl::InternalError(
+          fmt::format("metadata value is too long. value.length: {}", val.size()));
+    }
+    ProtobufWkt::Value protobuf_value;
+    protobuf_value.set_string_value(std::move(val));
+    return protobuf_value;
+  }
+};
+
+} // anonymous namespace
+
+Rule::Rule(const ProtoRule& rule) : rule_(rule) {
+  if (!rule_.has_on_present() && !rule_.has_on_missing()) {
+    throw EnvoyException("json to metadata filter: neither `on_present` nor `on_missing` set");
+  }
+
+  if (rule_.has_on_missing() && !rule_.on_missing().has_value()) {
+    throw EnvoyException(
+        "json to metadata filter: cannot specify on_missing rule with empty value");
+  }
+
+  if (rule_.has_on_error() && !rule_.on_error().has_value()) {
+    throw EnvoyException("json to metadata filter: cannot specify on_error rule with empty value");
+  }
+
+  // Support key selectors only.
+  for (const auto& selector : rule.selectors()) {
+    keys_.push_back(selector.key());
+  }
+}
+
+FilterConfig::FilterConfig(
+    const envoy::extensions::filters::http::json_to_metadata::v3::JsonToMetadata& proto_config,
+    Stats::Scope& scope)
+    : stats_{ALL_JSON_TO_METADATA_FILTER_STATS(POOL_COUNTER_PREFIX(scope, "json_to_metadata."))},
+      request_buffer_limit_bytes_(
+          PROTOBUF_GET_WRAPPED_OR_DEFAULT(proto_config, request_buffer_limit_bytes, 1024 * 1024)),
```

Comment:
note this is not using the underlying connection buffer value -- just the connection buffer size default of 1MB. e.g. say a deployment sets that buffer to 2MB, this would just be 1MB if unset. It should either be tied to that value or, I think (better) be independent of it.

For the latter that means that we can have this be some value independent of connection buffer, defaulting to using either 1MB or the HTTP filter chain's default sizes -- though this is protocol dependent.